src/Umbraco.Infrastructure/Runtime/AppPluginsManifestWatcherNotificationHandler.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Umbraco.Core;
+using Umbraco.Core.Events;
+using Umbraco.Core.Hosting;
+using Umbraco.Core.Manifest;
+
+namespace Umbraco.Infrastructure.Runtime
+{
+    /// <summary>
+    /// Starts monitoring AppPlugins directory during debug runs, to restart site when a plugin manifest changes.
+    /// </summary>
+    public sealed class AppPluginsManifestWatcherNotificationHandler : INotificationHandler<UmbracoApplicationStarting>
+    {
+        private readonly ManifestWatcher _manifestWatcher;
+        private readonly IHostingEnvironment _hostingEnvironment;
+
+        public AppPluginsManifestWatcherNotificationHandler(ManifestWatcher manifestWatcher, IHostingEnvironment hostingEnvironment)
+        {
+            _manifestWatcher = manifestWatcher ?? throw new ArgumentNullException(nameof(manifestWatcher));
+            _hostingEnvironment = hostingEnvironment ?? throw new ArgumentNullException(nameof(hostingEnvironment));
+        }
+
+        public Task HandleAsync(UmbracoApplicationStarting notification, CancellationToken cancellationToken)
+        {
+            if (!_hostingEnvironment.IsDebugMode)
+            {
+                return Task.CompletedTask;
+            }
+
+            var appPlugins = _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.AppPlugins);
+
+            if (!Directory.Exists(appPlugins))
```

Comment:
I don't understand why this is still here. The ManifestWatcher already does checks. 

https://github.com/umbraco/Umbraco-CMS/pull/9586/files#diff-fc7600f9983710ac030090cf5fe72eb528e0bc84c9fde3ba8eafe64426f46805R27

Reply:
Check the test failures, Directory.GetDirectories throws if we don’t check

Reply:
The whole method could be 

``` c#
if (!_hostingEnvironment.IsDebugMode)
{
    string appPlugins = _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.AppPlugins);
    _manifestWatcher.WatchDirectories(Directory.GetDirectories(appPlugins));
}

return Task.CompletedTask;
```

Reply:
The fix should be in the ManifestWatcher then.

Reply:
Manifest watcher should what take the parent path and add watchers for child directories?

Reply:
This would work in `ManifestWatcher` called via 

```c#
foreach (var packageFolder in ParseFolders(packageFolders))...
```

```c#
private static IEnumerable<string> ParseFolders(params string[] packageFolders)
{
    var valid = new List<string>();
    foreach (string folder in packageFolders)
    {
        if (File.Exists(Path.Combine(folder, "package.manifest")))
        {
            valid.Add(folder);
        }
        else if (Directory.Exists(folder))
        {
            valid.AddRange(Directory.GetDirectories(folder).Where(x => File.Exists(Path.Combine(x, "package.manifest"))));
        }
    }

    return valid;
}
```

Reply:
Double check my logic (_I haven't slept nearly enough_) but that should allow checking a folder and any children.

Reply:
The exception is thrown here in the handler

```
    _manifestWatcher.WatchDirectories(Directory.GetDirectories(appPlugins));
```

Reply:
You would change that line to 
```c#
_manifestWatcher.WatchDirectories(appPlugins);
```

Reply:
That would of course work.

It's a change to the interface of ManifestWatcher and outside of the scope of what I'm looking to achieve which is simply remove components and composers.

Reply:
Changing the method name already breaks the contract. You haven’t made a logic change, at worst it would mean a duplicate directory check.